In [21]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

data = pd.read_pickle("bank_09.pkl")

print(data.head())

print(data.shape)

print(data['marital'].value_counts().shape)

   age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes  
(11000, 17)
(3,)


Comenzamos identificandoo las variables categóricas nominales, categóricas ordinales y las variables numéricas.

Realizando un shape, obtenemos que hay 17 variables y 11000 instancias.

Llevamos a cabo un head para ver que valores tienen las diferentes variables (numérico, categórico ordinal o categórico nominal):
Variables numéricas: age, balance, day, duration, campaign, pdays, previous
Variables categóricas nominales: job, default, housing, loan, contact, marital, poutcome, deposit, month
Variables categóricas ordinales: education

In [22]:

lista=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome', 'deposit']
for item in lista:
    if data[item].value_counts().shape[0] > 10:
        print("Cardinalidad", item, "es", data[item].value_counts().shape[0])

Cardinalidad job es 12
Cardinalidad month es 12


Hacemos una lista en la que incluimos los nombres de todas las variables categóricas. Seguidamente, recorremos dicha lista, comprobando para cada ítem, si su cardinalidad es mayor que 10. De ser así, lo imprimimos.

In [ ]:
lista=['age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'deposit']
for item in lista:
    print("La columna '", item, "' tiene ", (data[item]=="unknown").sum(), "valores unknown")
    print("La columna '", item, "' tiene ", (data[item].isna()).sum(), "valores NULL \n")

La columna ' age ' tiene  0 valores unknown
La columna ' age ' tiene  0 valores NULL 

La columna ' job ' tiene  68 valores unknown
La columna ' job ' tiene  97 valores NULL 

La columna ' marital ' tiene  0 valores unknown
La columna ' marital ' tiene  282 valores NULL 

La columna ' education ' tiene  494 valores unknown
La columna ' education ' tiene  0 valores NULL 

La columna ' default ' tiene  0 valores unknown
La columna ' default ' tiene  0 valores NULL 

La columna ' balance ' tiene  0 valores unknown
La columna ' balance ' tiene  0 valores NULL 

La columna ' housing ' tiene  0 valores unknown
La columna ' housing ' tiene  0 valores NULL 

La columna ' loan ' tiene  0 valores unknown
La columna ' loan ' tiene  0 valores NULL 

La columna ' contact ' tiene  2309 valores unknown
La columna ' contact ' tiene  0 valores NULL 

La columna ' day ' tiene  0 valores unknown
La columna ' day ' tiene  0 valores NULL 

La columna ' month ' tiene  0 valores unknown
La columna ' month ' 

Hacemos un bucle para ver la cantidad de unknowns y NULL que hay en cada columna y obtenemos que las columnas que tienen valores faltantes son:
La columna ' job ' tiene  68 valores unknown
La columna ' job ' tiene  97 valores NULL 
La columna ' marital ' tiene  282 valores NULL 
La columna ' education ' tiene  494 valores unknown
La columna ' contact ' tiene  2309 valores unknown
La columna ' poutcome ' tiene  8205 valores unknown
Se observa que hay muchas filas en las que se desconocen valores o no están escritos ni siquiera

In [24]:
lista=['age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'deposit']
for item in lista:
    if data[item].value_counts().shape[0] == 1:
        print("Columa constante es:", item)

Hacemos un bucle para comprobar si hay columnas constantes (comprobando si el número de valores posibles que presentan es igual a 1) y obtenemos que no hay ninguna columna constante

In [25]:
lista=['age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'deposit']
for item in lista:
    if len(data[item].value_counts()) == data.shape[0]:
        print("Columa ID es:", item)

Hacemos un bucle para comprobar si hay columnas ID (comprobamos si el número de valores posibles que presentan es igual al número de instancias) y obtenemos que no hay ninguna columna ID

Se trata de un problema de clasificación puesto que la variable objetivo "deposit" es una variable categórica

In [26]:
print(data['deposit'].value_counts())

deposit
no     5780
yes    5220
Name: count, dtype: int64


Comprobamos la cantidad de valores que presenta cada clase para la variable "deposit" y podemos observar que no está desbalanceado puesto que hay un número parecido de valores para cada clase

In [27]:
print(data['pdays'].dtype)

print(data['pdays'].value_counts())

int64
pdays
-1      8203
 92      105
 182      88
 91       82
 181      79
        ... 
 651       1
 214       1
 24        1
 701       1
 118       1
Name: count, Length: 472, dtype: int64


La variable "pdays", como hemos mencionado anteriormente, es una variable numérica. Por otro lado, la distribución de valores está bastante descompensada puesto que la mayoría de las filas (alrededor de un 74%) tienen un valor igual a "-1", lo que significa que la mayoría de clientes no han sido contactados o esto es desconocido. Debido a que hay un total de 472 valores distintos y que "-1" representa un 74%, el modelo se vería muy afectado por dicho valor, por lo que es necesario preprocesamiento. Por otro lado, más arriba hemos visto también que "pdays" no tiene ningún valor "unknown", "NULL" o "NaN"

In [28]:
data['contactado_previamente'] = (data['pdays']!= -1).astype(int)
data['pdays'] = data['pdays'].replace(-1, np.nan)

print(data['contactado_previamente'].value_counts())

contactado_previamente
0    8203
1    2797
Name: count, dtype: int64


Preprocesado de la variable pdays:
Para el preprocesado identificamos que la variable pdays aporta dos tipos de información: si la persona ha sido contactada previamente y, en caso afirmativo, cuánto tiempo ha pasado desde ese contacto. Por ello, realizaremos dos operaciones.

En primer lugar, crearemos una nueva variable binaria, contactado_previamente, que tomará el valor 0 si el cliente no ha sido contactado previamente y 1 en caso contrario.
En segundo lugar, sustituiremos los valores -1 de la columna pdays por NaN en la propia columna pdays, ya que -1 no representa una cantidad real de días, sino ausencia de contacto previo o un valor no aplicable.

Con estos cambios obtenemos una representación más correcta de la información original. La variable contactado_previamente recoge si existió o no contacto previo, mientras que la modificación de pdays conserva únicamente los valores válidos de días para aquellos clientes que sí fueron contactados. De este modo, evitamos tratar -1 como un valor numérico ordinario.

In [31]:
filtro = data['poutcome'] == 'unknown'

print("previous cuando poutcome = unknown")
display(data.loc[filtro, 'previous'].value_counts())

print("pdays cuando poutcome = unknown")
display(data.loc[filtro, 'pdays'].value_counts())

previous cuando poutcome = unknown


previous
0    8203
2       1
1       1
Name: count, dtype: int64

pdays cuando poutcome = unknown


pdays
188.0    1
98.0     1
Name: count, dtype: int64

Se ha comprobado que la categoría unknown de la variable poutcome coincide prácticamente en todos los casos con previous = 0 y con pdays = Nan (persona no contactada, que tras el preprocesado pasó de -1 a Nan). En concreto, esto ocurre en 8203 de 8205 casos, por lo que unknown no parece representar un faltante aleatorio, sino una categoría informativa asociada a la ausencia de contactos o campañas previas.